// @contentful/contentful-slatejs-adapter/dist/lib/__test__/contentful-helpers.js
import { BLOCKS, } from '@contentful/rich-text-types';
export function document() {
    var content = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        content[_i] = arguments[_i];
    }
    return {
        data: {},
        nodeType: BLOCKS.DOCUMENT,
        content: content,
    };
}
export function block(nodeType) {
    var content = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        content[_i - 1] = arguments[_i];
    }
    return {
        nodeType: nodeType,
        content: content,
        data: {},
    };
}
export function inline(nodeType) {
    var content = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        content[_i - 1] = arguments[_i];
    }
    return {
        nodeType: nodeType,
        content: content,
        data: {},
    };
}
export function text(value) {
    var marks = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        marks[_i - 1] = arguments[_i];
    }
    return {
        nodeType: 'text',
        data: {},
        marks: marks,
        value: value,
    };
}
export function mark(type) {
    return {
        type: type,
    };
}
