// cobrowse-agent-sdk/webpack.config.mjs
// const path = require('path')
import path from 'path'
import { fileURLToPath } from 'url'

const basePath = path.dirname(fileURLToPath(import.meta.url))

const common = {
  resolve: {
    modules: ['node_modules', path.resolve(basePath, './')]
  },
  mode: 'production',
  devtool: 'source-map',
  module: {
    rules: [
      {
        test: /^(?!.*\.spec\.js$).*\.js$/,
        loader: 'babel-loader',
        exclude: /node_modules/
      }
    ]
  }
}

const umdBuild = {
  ...common,
  name: 'umd-build',
  entry: './src/index.js',
  output: {
    library: {
      name: 'CobrowseAPI',
      type: 'umd',
      export: 'default'
    },
    filename: 'cobrowse-agent-sdk.js',
    path: path.resolve(basePath, './dist/umd-browser')
  }
}

const esmBuild = {
  ...common,
  name: 'esm-build',
  entry: './src/index.js',
  output: {
    library: {
      type: 'module'
    },
    filename: 'index.js',
    path: path.resolve(basePath, './dist/esm-browser')
  },
  experiments: {
    outputModule: true
  }
}

export default [umdBuild, esmBuild]
