// babel-plugin-transform-react-router-optimize/distribution/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

exports.default = function (_ref) {
	var t = _ref.types;

	var namespace = 'react-router/lib';
	var importMap = {
		createRoutes: 'RouteUtils',
		locationShape: 'PropTypes',
		routerShape: 'PropTypes',
		formatPattern: 'PatternUtils'
	};

	var canReplace = function canReplace(_ref2) {
		var specifiers = _ref2.specifiers;

		return specifiers.length > 0 && specifiers.every(function (specifier) {
			return t.isImportSpecifier(specifier) && specifier.imported.name !== 'default';
		});
	};

	var replace = function replace(specifiers) {
		return specifiers.map(function (_ref3) {
			var local = _ref3.local;
			var imported = _ref3.imported;

			var mapped = importMap[imported.name];

			if (typeof mapped !== 'undefined') {
				return t.importDeclaration([t.importSpecifier(local, imported)], t.stringLiteral(namespace + '/' + mapped));
			}

			return t.importDeclaration([t.importDefaultSpecifier(local)], t.stringLiteral(namespace + '/' + imported.name));
		});
	};

	return {
		visitor: {
			ImportDeclaration: function ImportDeclaration(path) {
				if (path.node.source.value === 'react-router') {
					if (canReplace(path.node)) {
						path.replaceWithMultiple(replace(path.node.specifiers));
					}
				}
			}
		}
	};
};
